drivers: flash: microchip: Introduce QSPI Flash controller driver

LiBinSHA opened the pull request:
Add QSPI flash support for sama7g54-ek. 
Add qspi nodes (microchip_sama7g5_qspi ) to sama7g5 device tree.
Add  NOR flash support for microchip_sama7g5_qspi and Initial support for DMA based Microchip sama7g5 QSPI controller.

Tested samples\drivers\jesd216  on sama7g54-curiosity with qspi1 (PASSED)

*** Booting Zephyr OS build d1cdc53357e3 ***
qspi@e0810000: SFDP v 1.6 AP ff with 3 PH
PH0: ff00 rev 1.6: 16 DW @ 30
Summary of BFP content:
DTR Clocking not supported
Addressing: 3-Byte only
4-KiBy erase: uniform
Support QSPI XIP
Support 1-1-1
Support 1-1-2: instr 3Bh, 0 mode clocks, 8 waits
Support 1-1-4: instr 6Bh, 0 mode clocks, 8 waits
Support 1-2-2: instr BBh, 4 mode clocks, 0 waits
Support 1-4-4: instr EBh, 2 mode clocks, 4 waits
Support 4-4-4: instr 0Bh, 2 mode clocks, 4 waits
Flash density: 8388608 bytes
ET1: instr 20h for 4096 By; typ 19 ms, max 38 ms
ET2: instr D8h for 8192 By; typ 19 ms, max 38 ms
ET3: instr D8h for 32768 By; typ 19 ms, max 38 ms
Chip erase: typ 32 ms, max 64 ms
Byte program: type 48 + 4 * B us, max 96 + 8 * B us
Page program: typ 1024 us, max 2048 us
Page program size: 256 By
Suspend: B0h ; Resume: 30h
HOLD or RESET Disable: unsupported
QER: 5
0-4-4 Mode methods: entry 0xc ; exit 0x30
4-4-4 Mode sequences: enable 0x02 ; disable 0x9
Soft Reset and Rescue Sequence support: 0x30
Status Register 1 support: 0x70
size = <67108864>; // bits
sfdp-bfp = [
        fd 20 f1 ff  ff ff ff 03  44 eb 08 6b  08 3b 80 bb
        fe ff ff ff  ff ff 00 ff  ff ff 44 0b  0c 20 0d d8
        0f d8 10 d8  20 91 48 24  80 6f 1d 81  ed 0f 77 38
        30 b0 30 b0  f7 ff ff ff  29 c2 5c ff  f0 30 c0 80
        ];
PH1: ff81 rev 1.0: 6 DW @ 100
sfdp-ff81 = [
        ff 00 04 ff  f3 7f 00 00  f5 7f 00 00  f9 ff 7d 00
        f5 7f 00 00  f3 7f 00 00
        ];
PH2: 01bf rev 2.0: 28 DW @ 200
sfdp-01bf = [
        bf 26 43 ff  b9 5f fd ff  30 f2 60 f3  32 ff 0a 12
        23 46 ff 0f  19 32 0f 19  19 ff ff ff  ff ff ff ff
        00 66 99 38  ff 05 01 35  06 04 02 32  b0 30 72 42
        8d e8 98 88  a5 85 c0 9f  af 5a ff ff  06 ec 06 0c
        00 03 08 0b  ff ff ff ff  ff 07 ff ff  02 02 ff 06
        03 00 fd fd  04 07 00 fc  03 00 fe fe  02 02 07 0e
        30 04 91 62  42 ce 8e 40  00 1e c0 00  00 02 ce 8e
        ];
jedec-id = [26 43 bf];

Tested samples\drivers\spi_flash on sama7g54-curiosity with qspi1. (PASSED) 

ONE: AT91Bootstrap completed. Can load application via JTAG and jump.
*** Booting Zephyr OS build d1cdc53357e3 ***

qspi@e0810000 SPI flash testing
==========================
Perform test on single sector
Test 1: Flash erase
Flash erase succeeded!

Test 2: Flash write
Attempting to write 4 bytes
Data read matches data written. Good!!

LiBinSHA committed:
soc: microchip: sam: add qspi support for sama7g5

Add GCLK configurations to support qspi0 and qspi1.
Add MMU configurations for qspi peripherial.
Add MMU configurations for QSPI0 and QSPI1 memory space.

Signed-off-by: Li Bin <bin.li@microchip.com>

LiBinSHA committed:
drivers: flash: Introduce QSPI G1 flash driver.

Add support for DMA based Microchip sama7g5 QSPI controller.
Add NOR flash support for microchip_sama7g5_qspi device.
Reuses existing jesd216 library.

Signed-off-by: Li Bin <bin.li@microchip.com>

LiBinSHA committed:
boards: microchip: sam: add qspi device for sama7g5

Add qspi0 qspi1 for sama7g5-ek.

Signed-off-by: Li Bin <bin.li@microchip.com>

LiBinSHA committed:
samples: drivers: add samples for microchip qspi

Add jesd216 test case for microchip_sama7g5_qspi device (qspi1) on
sama7g54-curiosity board.
Add spi-flash test case for microchip_sama7g5_qspi device (qspi1) on
sama7g54-curiosity board.

Signed-off-by: Li Bin <bin.li@microchip.com>

nandojve commented on dts/bindings/flash_controller/microchip,qspi-controller.yaml:
```diff
@@ -0,0 +1,54 @@
+    type: int
+    default: 0x0
+    description: |
+      Flash Memory size in bits, required if skip-sfdp is enabled.
```
How to ensure that user will fill this option when `skip-sfdp;` ?

nandojve commented on boards/microchip/sam/sama7g54_ek/sama7g54_ek.dts:
```diff
@@ -408,3 +439,15 @@
+&qspi0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_qspi0_default>;
+	status = "disabled";
```
You can drop this line. By default the `qspi` is disabled at soc dts.

nandojve commented on drivers/flash/Kconfig.sam_qspi:
```diff
@@ -0,0 +1,12 @@
+	select FLASH_HAS_DRIVER_ENABLED
+	select FLASH_HAS_EXPLICIT_ERASE
+	select FLASH_JESD216
+	select DMA
```
The DMA driver is always selected but the usage is optional in devicetree.
Why customer would use the QSPI without DMA ?
Shall we "force" user to always define the DMA?
Can user select different DMA channels or those are fixed ?

nandojve commented on drivers/flash/flash_sam_ospic.c:
```diff
@@ -0,0 +1,389 @@
+	if (mode->cmd_buswidth != spi_flash_protocol_get_inst_nbits(op->proto)) {
+		return false;
+	}
+	if (mode->addr_buswidth != spi_flash_protocol_get_addr_nbits(op->proto)) {
```
Always add an empty line between distinct checks.
Fix this in all driver.

LiBinSHA replied:
Sure will update

nandojve commented on drivers/flash/flash_sam_ospic.c:
```diff
@@ -0,0 +1,389 @@
+
+static int qspi_reg_sync(qspi_registers_t *qspi)
+{
+	uint32_t timeout = 300;
```
Should we create a define with the timeout ?
BTW, why some parts use 100ms and here is 300ms ?

nandojve commented on drivers/flash/flash_sam_ospic.c:
```diff
@@ -0,0 +1,389 @@
+int qspi_sama7g5_init(struct qspi_priv *priv)
+{
+	qspi_registers_t *qspi;
+	uint32_t timeout = 100;
```
```diff
-uint32_t timeout = 100;
+uint32_t timeout_ms = 100;
```

nandojve reviewed (changes requested):
Hi @LiBinSHA ,

Nice work, some cosmetics and a few questions.
Make sure that CI is full green

LiBinSHA commented on dts/bindings/flash_controller/microchip,qspi-controller.yaml:
```diff
@@ -0,0 +1,54 @@
+
+compatible: "microchip,sama7g5-qspi"
+
+include: [flash-controller.yaml, pinctrl-device.yaml]
```
Sure will update

LiBinSHA commented on samples/drivers/jesd216/src/main.c:
```diff
@@ -312,20 +314,17 @@ int main(void)
 	} u;
 	const struct jesd216_sfdp_header *hp = &u.sfdp;
 	int rc = flash_sfdp_read(dev, 0, u.raw, sizeof(u.raw));
-
```
Sure will update

LiBinSHA commented on drivers/flash/flash_sam_ospic.c:
```diff
@@ -0,0 +1,389 @@
+	case QSPI_IFR_WIDTH_OCT_CMD:
+		mode_cycle_bits = 8;
+	default:
+		return -1;
```
Sure will update

nandojve commented on drivers/flash/flash_sam_ospic.c:
```diff
@@ -0,0 +1,420 @@
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
```
drop one empty line

nandojve commented on drivers/flash/flash_sam_qspi.h:
```diff
@@ -0,0 +1,121 @@
+ *
+ */
+
+/**
```
I think it is mandatory add guards in all header files.

```diff
 */
+
+#ifndef ZEPHYR_DRIVERS_FLASH_FLASH_SAM_QSPI_H_
+#define ZEPHYR_DRIVERS_FLASH_FLASH_SAM_QSPI_H_
+
/**
```

nandojve commented on drivers/flash/flash_sam_qspi.h:
```diff
@@ -0,0 +1,121 @@
+
+
+#define SFLASH_INST_ULBPR			0x98
+#define SFLASH_MAX_ID_LEN				6
```
```diff
#define SFLASH_INST_ULBPR			0x98
-#define SFLASH_MAX_ID_LEN				6
+#define SFLASH_MAX_ID_LEN			6
```

LiBinSHA replied:
Sure will update

LiBinSHA commented on dts/arm/microchip/sam/sama7g5.dtsi:
```diff
@@ -711,5 +711,26 @@
 			};
 		};
+
+		qspi0: qspi@e080c000 {
```
Sure will update

LiBinSHA commented on drivers/flash/flash_sam_qspi.h:
```diff
@@ -0,0 +1,121 @@
+};
+
+int qspi_sama7g5_init(struct qspi_priv *priv);
+int qspi_exec_op(struct qspi_priv *priv, const struct qspi_mem_op *op);
```
Sure will update

LiBinSHA commented on drivers/flash/flash_sam_qspi_nor.c:
```diff
@@ -0,0 +1,847 @@
+		}
+	}
+	/* No conversion found, keep input op code. */
+	return inst;
```
Sure will update

LiBinSHA commented on drivers/flash/flash_sam_qspi_nor.c:
```diff
@@ -0,0 +1,847 @@
+			      &flash_sam_qspi_config_##N,\
+			      POST_KERNEL,\
+			      CONFIG_FLASH_INIT_PRIORITY,\
+			      &flash_sam_nor_api);
```
Sure will update

nandojve commented on boards/m5stack/m5stack_nanoc6/board.yml:
This should not be part of this commit at least.
BTW, this should be a commit with mention "dts: ...."
Could this due to a rebase or a commit ammend mistake ?

LiBinSHA replied:
This is due to a re-base, i will update it.

nandojve commented on drivers/flash/flash_sam_qspi.h:
```diff
@@ -0,0 +1,126 @@
+int qspi_sama7g5_init(struct qspi_priv *priv);
+int qspi_exec_op(struct qspi_priv *priv, const struct qspi_mem_op *op);
+
+#endif /* ZEPHYR_DRIVERS_FLASH_FLASH_SAM_QSPI_H_ */
```
require new line at end of file

LiBinSHA committed:
dts: arm: microchip: sam: Add qspi devices to sama7g5

Add qspi0 and qspi1 devices to sama7g5.
Add microchip QSPI controller description for sama7g5.

Signed-off-by: Li Bin <bin.li@microchip.com>

nandojve commented on dts/bindings/flash_controller/microchip,qspi-controller.yaml:
```diff
@@ -0,0 +1,56 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+description: Microchip QSPI Flash controller
```
this commit MUST be the 2nd one, before driver.

JarmouniA commented on boards/microchip/sam/sama7g54_ek/sama7g54_ek.dts:
```diff
@@ -418,3 +449,18 @@
 	status = "okay";
 };
+
+&qspi0 {
```
missing a child node that represents the flash device.
ex. https://github.com/zephyrproject-rtos/zephyr/blob/6c410584dd64b5e51540ce931342c315ccbb5d9b/boards/st/stm32h750b_dk/stm32h750b_dk-common.dtsi#L196

LiBinSHA replied:
The `child_node` described by `ext_flash` is optional. For the microchip_sama7g5_qspi driver, by default, it obtains all SPI NOR parameters from the JESD216 specification (SFDP), which defines methods for describing the functional capabilities and characteristics of serial flash devices.

If the “skip-sfdp” option is enabled (indicating the device should skip SFDP parsing to obtain flash parameters), the driver will force SPI mode.

LiBinSHA replied:
Yes, there are currently no plans to use the expanded flash memory for XIP. I believe its sole purpose is to increase flash capacity for MCU apps. For us (MPU), SD/MMC remains the most mainstream option.

nandojve commented on dts/bindings/flash_controller/microchip,qspi-controller.yaml:
```diff
@@ -0,0 +1,56 @@
+# Copyright (c) 2025 Microchip Technology  Inc. and its subsidiaries
```
Update all new files with the correct date.
```diff
-# Copyright (c) 2025 Microchip Technology  Inc. and its subsidiaries
+# Copyright (c) 2025-2026 Microchip Technology  Inc. and its subsidiaries
```

LiBinSHA replied:
Sure will update

nandojve commented on drivers/flash/flash_sam_qspi_nor.c:
```diff
@@ -0,0 +1,851 @@
+
+		if (data->addr_len == 4) {
+			data->write_inst = SPI_NOR_CMD_PP_4B;
+		}
```
```diff
		rc = qspi_quad_enable(dev);
		if (rc < 0) {
			LOG_ERR("Failed to enable Quad mode: %d\n", rc);
			return rc;
		}
+
		data->write_inst = SPI_NOR_CMD_PP;

		if (data->addr_len == 4) {
			data->write_inst = SPI_NOR_CMD_PP_4B;
		}
```

LiBinSHA replied:
Sure will update

LiBinSHA commented on drivers/flash/flash_sam_qspi_nor.c:
```diff
@@ -0,0 +1,851 @@
+	/* write non-zero size */
+	if (size == 0) {
+		return 0;
+	}
```
Sure will update

LiBinSHA commented:
Thanks for your comments.
I Update the CONFIG_ARM_MMU_NUM_L2_TABLES to 320 in samples\drivers\spi_flash\boards\sama7g54_ek.conf.

LiBinSHA commented on dts/arm/microchip/sam/sama7/sama7g5/sama7g5.dtsi:
```diff
@@ -126,6 +126,24 @@
 			clocks = <&clk32k 1>;
 		};
 
+		qspi0: flash-controller@e080c000 {
```
sure, updated

LiBinSHA commented on drivers/flash/CMakeLists.txt:
```diff
@@ -48,6 +48,7 @@ zephyr_library_sources_ifdef(CONFIG_FLASH_RENESAS_RZ_QSPI_SPIBSC flash_renesas_r
 zephyr_library_sources_ifdef(CONFIG_FLASH_RENESAS_RZ_QSPI_XSPI flash_renesas_rz_qspi.c)
 zephyr_library_sources_ifdef(CONFIG_FLASH_RPI_PICO flash_rpi_pico.c)
 zephyr_library_sources_ifdef(CONFIG_FLASH_S3AXX04 flash_s3axx04.c)
+zephyr_library_sources_ifdef(CONFIG_FLASH_SAM_QSPI flash_sam_qspi_nor.c flash_sam_ospic.c)
```
sure, updated

LiBinSHA commented on dts/bindings/flash_controller/microchip,qspi-controller.yaml:
```diff
@@ -0,0 +1,56 @@
+
+description: Microchip QSPI Flash controller
+
+compatible: "microchip,sama7g5-qspi"
```
sure, updated

LiBinSHA replied:
Reworked, update to 'sama7g5-qspi-flash' in dtsi, drivers, and samples.

LiBinSHA commented on dts/bindings/flash_controller/microchip,qspi-controller.yaml:
```diff
@@ -0,0 +1,56 @@
+include:
+ - flash-controller.yaml
+ - pinctrl-device.yaml
+
```
sure, updated

LiBinSHA commented:
I also got problem after rebase and update Python to 3.12..
This is error message when build hello_world:

user@user-Latitude-E6420:~/zephyrproject/zephyr$ west build -p always -b sama7g54_ek samples/hello_world/
-- west build: making build dir /home/user/zephyrproject/zephyr/build pristine
-- west build: generating a build system
Loading Zephyr default modules (Zephyr base).
-- Application: /home/user/zephyrproject/zephyr/samples/hello_world
-- CMake version: 3.28.3
-- Found Python3: /usr/bin/python3 (found suitable version "3.12.3", minimum required is "3.12") found components: Interpreter
-- Cache files will be written to: /home/user/.cache/zephyr
-- Zephyr version: 4.3.99 (/home/user/zephyrproject/zephyr)
-- Found west (found suitable version "1.5.0", minimum required is "0.14.0")
-- Board: sama7g54_ek, qualifiers: sama7g54
-- Found host-tools: zephyr 0.17.0 (/home/user/zephyrproject/zephyr-sdk-0.17.0)
-- Found toolchain: zephyr 0.17.0 (/home/user/zephyrproject/zephyr-sdk-0.17.0)
-- Found Dtc: /usr/bin/dtc (found suitable version "1.7.0", minimum required is "1.4.6")
-- Found BOARD.dts: /home/user/zephyrproject/zephyr/boards/microchip/sam/sama7g54_ek/sama7g54_ek.dts
-- Generated zephyr.dts: /home/user/zephyrproject/zephyr/build/zephyr/zephyr.dts
-- Generated pickled edt: /home/user/zephyrproject/zephyr/build/zephyr/edt.pickle
-- Generated devicetree_generated.h: /home/user/zephyrproject/zephyr/build/zephyr/include/generated/zephyr/devicetree_generated.h
Parsing /home/user/zephyrproject/zephyr/Kconfig
Loaded configuration '/home/user/zephyrproject/zephyr/boards/microchip/sam/sama7g54_ek/sama7g54_ek_defconfig'
Merged configuration '/home/user/zephyrproject/zephyr/samples/hello_world/prj.conf'
Configuration saved to '/home/user/zephyrproject/zephyr/build/zephyr/.config'
Kconfig header saved to '/home/user/zephyrproject/zephyr/build/zephyr/include/generated/zephyr/autoconf.h'
-- Found GnuLd: /home/user/zephyrproject/zephyr-sdk-0.17.0/arm-zephyr-eabi/arm-zephyr-eabi/bin/ld.bfd (found version "2.38")
-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- The ASM compiler identification is GNU
-- Found assembler: /home/user/zephyrproject/zephyr-sdk-0.17.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc
-- Found gen_kobject_list: GEN_KOBJECT_LIST-NOTFOUND
-- Configuring done (9.4s)
-- Generating done (0.1s)
-- Build files have been written to: /home/user/zephyrproject/zephyr/build
-- west build: building application
ninja: error: 'zephyr/GEN_KOBJECT_LIST-NOTFOUND', needed by 'zephyr/include/generated/zephyr/driver-validation.h', missing and no known rule to make it
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build /home/user/zephyrproject/zephyr/build

LiBinSHA commented:
Fixed now

nandojve reviewed (dismissed):
Ping @de-nordic

de-nordic commented on drivers/flash/flash_sam_qspi_nor.c:
```diff
@@ -0,0 +1,890 @@
+	/* read non-zero size */
+	if (size == 0) {
+		return 0;
+	}
```
Move above `qspi_address_is_valid`, if you are not going to do any work, do not fail API call due to incorrect parameters.

LiBinSHA replied:
Updated

nandojve commented on dts/bindings/flash_controller/microchip,sama7g5-qspi-flash.yaml:
```diff
@@ -0,0 +1,58 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+description: Microchip QSPI Flash controller
```
Add example entry:
https://github.com/zephyrproject-rtos/zephyr/blob/7bbe9da0892c2c3bc2cdefbc5e60b941f9a36da4/dts/bindings/usb/atmel%2Csam-udp.yaml#L21-L40

LiBinSHA replied:
sure, updated.

LiBinSHA commented on drivers/flash/flash_sam_qspi_nor.c:
```diff
@@ -0,0 +1,858 @@
+
+	erase_op.addr.nbytes = data->addr_len;
+	erase_op.proto = SFLASH_PROTO_1_1_1;
+
```
Updated

LiBinSHA commented on drivers/flash/flash_sam_qspi_nor.c:
```diff
@@ -0,0 +1,858 @@
+	erase_op.proto = SFLASH_PROTO_1_1_1;
+
+	/* erase area must be subregion of device */
+	if ((addr < 0) || ((size + addr) > flash_size)) {
```
Updated

LiBinSHA commented on drivers/flash/flash_sam_qspi_nor.c:
```diff
@@ -0,0 +1,858 @@
+	if (size == 0) {
+		return 0;
+	}
+
```
Updated

LiBinSHA commented on drivers/flash/flash_sam_qspi_nor.c:
```diff
@@ -0,0 +1,890 @@
+	struct flash_sam_nor_data *data = dev->data;
+	size_t flash_size = data->size;
+
+	return (addr >= 0) && ((uint64_t)addr + (uint64_t)size <= flash_size);
```
Updated

LiBinSHA commented on drivers/flash/flash_sam_qspi_nor.c:
```diff
@@ -0,0 +1,858 @@
+	struct qspi_priv hqspi = {0};
+	struct qspi_mem_op read_op = {0};
+	int ret;
+
```
Updated

nandojve commented:
ping @de-nordic

cssedev commented on drivers/flash/flash_sam_ospic.c:
```diff
@@ -0,0 +1,451 @@
+	struct dma_status stat;
+	uint32_t channel;
+	static struct dma_config dma_cfg = {0};
+	static struct dma_block_config dma_block_cfg;
```
Should this and the `struct dma_config dma_cfg` above not be allocated per-device instead of static?

nandojve replied:
Agreed!

cssedev commented on drivers/flash/flash_sam_ospic.c:
```diff
@@ -0,0 +1,451 @@
+	if (op->data.dir == QSPI_MEM_DATA_IN) {
+		qspi->QSPI_RICR = icr;
+	} else {
+		qspi->QSPI_WICR = icr;
```
The `QSPI_WICR` register is written with a value prepared using macros/defines specific for the `QSPI_RICR` register. It is not an issue for the SAMA7G54, as the two registers have the same layout, but it may have implications down the line.

nandojve replied:
I don't see a particular issue at moment, but are you mention something like maybe we pass by the mask before set the register?

cssedev replied:
The `icr` variable is initialized further up with `icr = QSPI_RICR_RDINST(op->cmd.opcode);`, so a define meant for the `RICR` register can inadvertently get used to set the value of the `WICR` register instead. It is not an issue for the SAMA7G54, and it may be that for all implementations of the peripheral IP they will be the same, I just find it a bit odd.
If I attempt to "fix" it by splitting the initial assignment of the `icr` variable as well, I just get a warning that both paths have the same code, so there may not be a clear cut solution for this.

cssedev commented on soc/microchip/sam/sama7/sama7g5/soc.c:
```diff
@@ -92,6 +92,22 @@ static const struct arm_mmu_region mmu_regions[] = {
 		   (MMU_REGION_FLAT_ENTRY("pwm", PWM_BASE_ADDRESS, 0x500,
 					  MT_STRONGLY_ORDERED | MPERM_R | MPERM_W),))
 
+	IF_ENABLED(DT_NODE_HAS_STATUS_OKAY(DT_NODELABEL(qspi0)),
```
In the driver initialization (`FLASH_SAM_QSPI`), the QSPI peripheral base and memory base is read from the device tree, but here they are fixed from defines. Is it possible to read them from the device tree here instead?

Also, the entire memory range is mapped for the `qspi0mem` and `qspi1mem`. If the flash memory is significantly smaller, wouldn't it make sense to read this from the device tree properties as well? This would also limit the number of additional L2 pages necessary in the `CONFIG_ARM_MMU_NUM_L2_TABLES`.

nandojve replied:
> Is it possible to read them from the device tree here instead?

yes, agreed!

> If the flash memory is significantly smaller, wouldn't it make sense to read this from the device tree properties as well? This would also limit the number of additional L2 pages necessary in the CONFIG_ARM_MMU_NUM_L2_TABLES.

Could you give more details about which flash you are mentioning?

cssedev replied:
The memory area for `qspi0mem` and `qspi1mem` on the SAMA7G54 is 256 MB large. In my case, I only have a 2 MB external flash connected. I would rather have a smaller area mapped if possible, such that I get a hard fault from the MMU if I attempt to perform an operation of more than 2 MB.

cssedev commented on soc/microchip/sam/sama7/sama7g5/soc.c:
```diff
@@ -200,4 +215,18 @@ void soc_early_init_hook(void)
+
+	/* Enable generic clock for QSPI0, frequency is 50MHz */
+	if (DT_NODE_HAS_STATUS_OKAY(DT_NODELABEL(qspi0))) {
+		PMC_REGS->PMC_PCR = PMC_PCR_CMD(1) | PMC_PCR_GCLKEN(1) | PMC_PCR_EN(1) |
```
The peripheral and generic clocks are enabled here, but also in `flash_sam_nor_init`. I assume it is due to the generic clock, which doesn't have an API via the clock API as far as I am aware?

nandojve replied:
Correct! At moment, not all infra for clocks in these Cortex-A are in place. We adopted to bring and make them available. Improvements will be made in future.

cssedev reviewed (commented):
@nandojve I am unable to give a proper review in the context of Zephyr, as I have limited experience with all the intricacies of the system. I have tried to highlight the sections I find odd.

* In general, when using register defines the `X_Msk` option is often used when setting a single-bit flag. I would go with `X_1` instead (where `X_1` maps to the flag being set of course). An example of this would be to use `QSPI_IFR_INSTEN_1` instead of `QSPI_IFR_INSTEN_Msk`.
* It likely has no real performance impact, but there is no support for quad write operations. All write operations are mapped to 1-1-1 operations.
*  The driver heavily relies on polling of the status registers of the QSPI peripheral and the DMA peripheral. At least for the DMA, this could be handled via ISR callback.
* Clock configuration in general for the SAMA7G54 seems locked to some default values and not configurable. Here, the configuration appears to be locked to a 50 MHz clock. Ideally I would assume this could be handled in the device tree.

A small performance note, when comparing with the official Microchip SAM-BA tool and the SoC built-in ROMboot. These numbers are from an earlier version of the driver, but they may still be of interest. When erasing the entire external flash chip it takes ~15 seconds. Programming the entire external flash takes ~20 seconds. The driver here significantly improves erase performance, since it can perform a full chip erase instead of sector by sector. The write step takes upwards of ~85 seconds. If I replace all the polling with busy-wait operations instead, I can get write performance comparable to the SAM-BA tool, i.e. less than 20 seconds for the same binary image. I do not know if these busy-wait operations can be applied in general, as it will impact the responsiveness of the system.

nandojve commented:
Hi @cssedev ,

Thank you for your valuable input.

> In general, when using register defines the X_Msk option is often used when setting a single-bit flag. I would go with X_1 instead (where X_1 maps to the flag being set of course). An example of this would be to use QSPI_IFR_INSTEN_1 instead of QSPI_IFR_INSTEN_Msk.

There is no specific "size" in the foo_Msk semantics. It can be 1 bit or multiple. In this case, there is no enforcement from my side. One think that maybe we left behind is to make sure the registers bits are safe when do the attribution. Maybe we could re-evaluate the mask to make sure we are not affecting other bits. But then again, sometimes registers having clear/set and which already filter correct bits. So, to do fully review this properly would be only with a datasheet always open.

The only way to make it clear to any public could be `force the mask evaluation` at attribution as a rule. CC: @NhMchp 

About optimizations, from my side I don't pick in this things because I believe it is more important have the feature. I think, optimization is a natural thing that came with real need and use cases. Linux has been optimized over more than 30 years and it never finish : ) and in Zephyr will not be different.

@LiBinSHA ,

I think @cssedev brought interesting things some I agreed that are necessary now. Optimization I let to you decide.

cssedev reviewed (commented):
With the caveat that I have not been able to test this version on real hardware, the only thing I really would like to see changed is the configuration of the MMU tables in the SoC setup file for the SAMA7G54. These still use values from defines instead of the values entered in the device tree. I assume something along these lines would work:

```
    IF_ENABLED(DT_NODE_HAS_STATUS_OKAY(DT_NODELABEL(qspi0)),
        (MMU_REGION_FLAT_ENTRY("qspi0",
            DT_REG_ADDR_BY_IDX(DT_NODELABEL(qspi0), 0),
            DT_REG_SIZE_BY_IDX(DT_NODELABEL(qspi0), 0),
            MT_STRONGLY_ORDERED  | MPERM_R | MPERM_W),))

    IF_ENABLED(DT_NODE_HAS_STATUS_OKAY(DT_NODELABEL(qspi0)),
        (MMU_REGION_FLAT_ENTRY("qspi0mem",
            DT_REG_ADDR_BY_IDX(DT_NODELABEL(qspi0), 1),
            DT_REG_SIZE_BY_IDX(DT_NODELABEL(qspi0), 1),
            MT_STRONGLY_ORDERED  | MPERM_R | MPERM_W),))
```

It may be that it could even be simplified further by using some of the logic from the FlexCOM/MCAN/XDMAC defines

```
#define MMU_REGION_QSPI_DEFN(idx, n)									\
		IF_ENABLED(DT_NODE_HAS_STATUS_OKAY(DT_NODELABEL(qspi##n)),		\
			(MMU_REGION_FLAT_ENTRY("qspi"#n,							\
				DT_REG_ADDR_BY_IDX(DT_NODELABEL(qspi##n), 0),			\
				DT_REG_SIZE_BY_IDX(DT_NODELABEL(qspi##n), 0),			\
				MT_STRONGLY_ORDERED | MPERM_R | MPERM_W),				\
			MMU_REGION_FLAT_ENTRY("qspi"#n"mem",						\
				DT_REG_ADDR_BY_IDX(DT_NODELABEL(qspi##n), 1),			\
				DT_REG_SIZE_BY_IDX(DT_NODELABEL(qspi##n), 1),			\
				MT_STRONGLY_ORDERED | MPERM_R | MPERM_W),				\
			)															\
		)

FOR_EACH_IDX(MMU_REGION_QSPI_DEFN, (), 0, 1)
```

I am not aware of the requirements to get this to work with XiP. It may be that it is enough to just add `MPERM_X` to the memory section. Otherwise a virtual entry could be added in order to mark the region as `MPERM_R | MPERM_X`. It is not a priority for us.

The only other comment I would flag is the usage of the `RICR` register defines potentially being used for the `WICR`. If the peripheral IP never changes, it is fine.

```
icr = QSPI_RICR_RDINST(op->cmd.opcode);

...

if (op->data.dir == QSPI_MEM_DATA_IN) {
	qspi->QSPI_RICR = icr;
} else {
	qspi->QSPI_WICR = icr;
}
```

LiBinSHA commented:
Hello, cssedev,

Your comments on SOC were very insightful. I've updated.
Regarding the use of RICR, the peripheral IP never changes.

Rest Regards,
LiBin

cssedev commented:
Hello @LiBinSHA,

the last change I can see for the SAMA7G54 MMU setup in the `soc.c` file is in commit 203ee1b8203adb6bc94cde434ddafdba056efffc, which seems to be with fixed addresses and sizes from defines and not the device tree. Am I missing a commit somewhere, or are we talking past each other?

(note: I am out of office next week, and may be slow to reply)

LiBinSHA commented:
Sure, updated.

nandojve commented:
Hi @cssedev , @de-nordic ,
I was wondering if you still have anything open in this topic.

nandojve commented:
ping @de-nordic

LiBinSHA commented:
ping @de-nordic

christian-steiniche commented:
There is a discrepancy between the devicetree binding and the driver implementation.

1. **Binding**: The size property is explicitly defined as being in bits (consistent with jedec,jesd216.yaml)
2. **Implementation**: In flash_sam_qspi_nor.c, when skip-sfdp is enabled, the driver takes the size value from the devicetree and uses it directly as bytes (see line 780 and 873).

As a result, e.g. the value returned through the standard flash_api_get_size function is off by a factor of 8.

nandojve commented:
@LiBinSHA , can you check last comment?

LiBinSHA commented:
Yes, the size property is explicitly defined as being in bits  (consistent with jedec,jesd216.yaml), we got the size by parsing SFDP (serial flash discovery parameter), so we divide it by 8 to get the size in bytes. （ in function spi_nor_process_bfp() )
const size_t flash_size = (jesd216_bfp_density(bfp) / 8U);

When skip-sfdp is enabled, we get the flash size in byte directly from the device tree.

Best Regards,
LiBin

LiBinSHA commented:
@christian-steiniche

LiBinSHA commented:
@de-nordic, could you please merge this branch? It's been sitting there for too long...

nandojve commented:
Hi @LiBinSHA 
No merge commits are allowed. You need to rebase your commits on top of latest main and force a push.

LiBinSHA commented:
Hi @nandojve 
Sure, updated.

LiBinSHA commented:
@nandojve @cssedev @de-nordic 

Any update? Could you please merge this branch? It's been sitting there for too long...

LiBin

nandojve commented:
ping @de-nordic

de-nordic commented on drivers/flash/flash_sam_qspi_nor.c:
```diff
@@ -0,0 +1,900 @@
+	int ret = 0;
+	struct qspi_mem_op erase_op = {0};
+
+	if (!qspi_address_is_valid(dev, addr, size)) {
```
This should be preceded by size check and return if the value is zero.
The check can not be done in qspi_address_is_valid, because size is a valid non-error case, and here you will translate 0 to `-EINVA`L even for  size = 0.

LiBinSHA replied:
Thanks for pointing this out.
I've added the size-zero check before the address validation, this issue has been fixed.

de-nordic commented on drivers/flash/flash_sam_qspi_nor.c:
```diff
@@ -0,0 +1,900 @@
+	int ret = 0;
+	size_t page_size = data->page_size;
+
+	if (!qspi_address_is_valid(dev, addr, size)) {
```
This should be preceded by size check and return if the value is zero.
The check can not be done in qspi_address_is_valid, because size is a valid non-error case, and here you will translate 0 to `-EINVA`L even for  size = 0.

LiBinSHA replied:
Thanks for pointing this out.
I've added the size-zero check before the address validation, this issue has been fixed.

de-nordic commented on drivers/flash/flash_sam_qspi_nor.c:
```diff
@@ -0,0 +1,900 @@
+	struct qspi_mem_op read_op = {0};
+	int ret;
+
+	if (!qspi_address_is_valid(dev, addr, size)) {
```
This should be preceded by size check and return if the value is zero.
The check can not be done in qspi_address_is_valid, because size is a valid non-error case, and here you will translate 0 to `-EINVA`L even for  size = 0.

LiBinSHA replied:
Thanks for pointing this out.
I've added the size-zero check before the address validation, this issue has been fixed.

nandojve commented:
@LiBinSHA ,

need to fix CI, not sure if this is related to this PR too.

kartben commented:
> @LiBinSHA ,
> 
> need to fix CI, not sure if this is related to this PR too.

Rebasing on main should help

nandojve commented:
Hi @LiBinSHA ,
Need fix the compliance check.
